include/can.h

```diff
@@ -0,0 +1,369 @@
+/**
+ * @file
+ *
+ * @brief Public APIs for the CAN drivers.
+ */
+
+/*
+ * Copyright (c) 2018 Alexander Wachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __DRIVERS_CAN_H
+#define __DRIVERS_CAN_H
+
+/**
+ * @brief CAN Interface
+ * @defgroup can_interface CAN Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <zephyr/types.h>
+#include <device.h>
+#include <string.h>
+
+#define CAN_EX_ID      (1 << 31)
+#define CAN_MAX_STD_ID (0x7FF)
+#define CAN_STD_ID_MASK CAN_MAX_STD_ID
+#define CAN_EXT_ID_MASK (0x1FFFFFFF)
+#define CAN_MAX_DLC    (8)
+
+/* CAN_TX_* are the error flags from tx_callback and send.*/
+/** send successfully */
+#define CAN_TX_OK       (0)
+/** general send error */
+#define CAN_TX_ERR      (1)
+/** bus arbitration lost during sending */
+#define CAN_TX_ARB_LOST (2)
+/** controller is in bus off state */
+#define CAN_TX_BUS_OFF  (3)
+/** unexpected error */
+#define CAN_TX_UNKNOWN  (4)
+
+/** attach_* failed because there is no unused filter left*/
+#define CAN_NO_FREE_FILTER (-1)
+
+/** operation timed out*/
+#define CAN_TIMEOUT (1)
+
+/**
+ * @brief Statically define and initialize a can message queue.
+ *
+ * The message queue's ring buffer contains space for @a size messages.
+ *
+ * @param name Name of the message queue.
+ * @param size Number of can messages.
+ */
+#define CAN_DEFINE_MSGQ(name, size) K_MSGQ_DEFINE(name, sizeof(struct can_msg), size, 4)
+
+/**
+ * @typedef can_ide_t
+ * @brief Define if the message has a standard (11bit) or extended (29bit)
+ * identifier
+ */
+typedef enum {
+	can_standard_identifier,
+	can_extended_identifier
+} can_ide_t;
+
+/**
+ * @typedef can_rtr_t
+ * @brief Define if the message is a data or remote frame
+ */
+typedef enum {
+	can_dataframe,
+	can_remoterequest
+} can_rtr_t;
+
+/**
+ * @typedef can_mode_t
+ * @brief Defines the mode of the can controller
+ */
+typedef enum {
+	/*Normal mode*/
+	can_mode_normal,
+	/*Controller is not allowed to send dominant bits*/
+	can_mode_silent,
+	/*Controller is in loopback mode (receive own messages)*/
+	can_mode_loopback,
+	/*Combination of loopback and silent*/
+	can_mode_silent_loopback
+} can_mode_t;
+
+/**
+ * @brief can message structure
+ *
+ * Used to pass can messages from userspace to the driver and
+ * from driver to userspace
+ *
+ */
+struct can_msg {
+	/** Indicates the identifier type (standard or extended)
+	 * use can_ide_t for assignment
+	 */
+	u32_t id_type : 1;
```

Comment:
Using bitfields in structs that are received from wire is problematic as this is compiler specific stuff how it places the bitfields. I think better would be to not to use them here but have access functions that get/set those bits instead.

Reply:
Bitfields are here used to save memory. They are anyway not written directly form the controller registers but set with the enum.

Reply:
Ok. It just looked that this struct is directly mapped to bytes received from wire, especially as there was the __packed attribute there (which is usually not used for normal memory only structs).